libusb/os/events_posix.c

```diff
@@ -37,26 +37,21 @@
  *
  * Therefore use a custom event system based on browser event emitters. */
 #include <emscripten.h>
+#include <emscripten/atomic.h>
+#include <emscripten/threading.h>
 
-EM_JS(void, em_libusb_notify, (void), {
-	dispatchEvent(new Event("em-libusb"));
+EM_ASYNC_JS(int, em_libusb_wait_async, (const int* ptr, int expected_value, int timeout), {
+	return (await Atomics.waitAsync(HEAP32, ptr >> 2, expected_value, timeout).value) === 'ok';
 });
 
-EM_ASYNC_JS(int, em_libusb_wait, (int timeout), {
-	let onEvent, timeoutId;
-
-	try {
-		return await new Promise(resolve => {
-			onEvent = () => resolve(0);
-			addEventListener('em-libusb', onEvent);
-
-			timeoutId = setTimeout(resolve, timeout, -1);
-		});
-	} finally {
-		removeEventListener('em-libusb', onEvent);
-		clearTimeout(timeoutId);
+static int em_libusb_wait(int *ptr, int expected_value, int timeout)
+{
+	if (emscripten_is_main_runtime_thread()) {
+		return em_libusb_wait_async(ptr, expected_value, timeout);
+	} else {
+		return emscripten_atomic_wait_u32(ptr, expected_value, timeout) == ATOMICS_WAIT_OK;
```

Comment:
`emscripten_atomic_wait_u32` uses nanosecond units, `timeout` is milliseconds here.

Reply:
Ohh good catch, must have missed it when switching from JS-based API. It's pretty unfortunate that JS `Atomics.wait` and Wasm `atomics.wait` take different units 🤦🏼‍♂️

Reply:
```suggestion
		return emscripten_atomic_wait_u32(ptr, expected_value, 1'000'000LL * timeout) == ATOMICS_WAIT_OK;
```